prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -179,6 +192,121 @@ def filter_by_connected_components(lines, substations):
     return remaining_lines, remaining_substations
 
 
+def augment_line_voltages(
+    lines, substations, volt_class_defaults=None, state_threshold_100_161=0.95
+):
+    """Fill in voltages for lines with missing voltages, using a series of heuristics.
+    The ``lines`` dataframe will be modified in-place.
+
+    :param pandas.DataFrame lines: data frame of lines.
+    :param pandas.DataFrame substations: data frame of substations.
+    :param dict/pandas.Series volt_class_defaults: mapping of volt classes to default
+        replacement voltage values. If None, internally-defined defaults will be used.
+    :param int/float state_threshold_100_161: fraction of lines in the 100-161 kV range
+        which must be the same voltage for the most common voltage to be applied to
+        missing voltage lines in this state.
+    """
+
+    # Interpret input parameters
+    if volt_class_defaults is None:
+        volt_class_defaults = {
+            "UNDER 100": 69,
+            "220-287": 230,
+            "345": 345,
+            "500": 500,
+        }
+
+    # Replace dummy data with explicit 'missing'
+    lines.loc[lines.VOLTAGE == -999999, "VOLTAGE"] = pd.NA
+
+    # Set voltages based on voltage class defaults
+    null_voltages = lines.loc[lines.VOLTAGE.isna()]
+    replacement_voltages = null_voltages.VOLT_CLASS.map(volt_class_defaults)
+    lines.loc[lines.VOLTAGE.isna(), "VOLTAGE"] = replacement_voltages
+
+    # Use commonly-assigned voltages by state to fill 100-161 kV range
+    # Assume state grouping on one end is representative
+    missing_mask = lines.VOLTAGE.isna()
+    class_100_161_mask = lines.VOLT_CLASS == "100-161"
+    lines["STATE_1"] = lines.SUB_1_ID.map(substations.STATE)
+    state_100_161_lines = {
+        state: group.query("100 <= VOLTAGE <= 161").value_counts("VOLTAGE")
+        for state, group in lines.groupby("STATE_1")
+    }
+    most_common_100_161_voltage = {
+        state: voltages.idxmax()
+        if voltages.max() > state_threshold_100_161 * voltages.sum()
+        else None
+        for state, voltages in state_100_161_lines.items()
+    }
+    lines.loc[missing_mask & class_100_161_mask, "VOLTAGE"] = lines.loc[
+        missing_mask & class_100_161_mask
+    ].STATE_1.map(most_common_100_161_voltage)
+    lines.drop("STATE_1", axis=1, inplace=True)
+
+    # Create a mapping of lines to neighboring lines
+    sub_1_groupby = lines.groupby("SUB_1_ID")
+    sub_2_groupby = lines.groupby("SUB_2_ID")
+    empty_df = pd.DataFrame()
+    neighbors = lines.loc[lines.VOLTAGE.isna()].apply(
+        lambda x: set().union(
+            safe_get_group(sub_1_groupby, x.SUB_1_ID, empty_df).index,
+            safe_get_group(sub_1_groupby, x.SUB_2_ID, empty_df).index,
+            safe_get_group(sub_2_groupby, x.SUB_1_ID, empty_df).index,
+            safe_get_group(sub_2_groupby, x.SUB_2_ID, empty_df).index,
+        ),
+        axis=1,
+    )
```

Comment:
Maybe we could do this in one shot:
```python
sub1_lines = lines.groupby("SUB_1_ID")["ID"].apply(set)
sub2_lines = lines.groupby("SUB_2_ID")["ID"].apply(set)
sub_lines = sub1_lines.combine(sub2_lines, set.union, fill_value=set())
neighbors = lines.loc[lines.VOLTAGE.isna()].apply(lambda x: sub_lines.loc[x.SUB_1_ID]|sub_lines.loc[x.SUB_2_ID])
```

Reply:
`combine` is clever here, I haven't used that before. I'll play around with this. This comment also makes me realize that we probably want to use the `ID` field of `lines` as an index, which will require a bit of refactoring of those first two lines, but shouldn't be too bad.

Reply:
Done. This also let me delete the `safe_get_group` function I had added, which always felt kludgey.